pkg/scm/git/git.go

```diff
@@ -217,83 +219,95 @@ func makePathAbsolute(source string) string {
 // expect git clone spec syntax; it also provides details if the file://
 // proto was explicitly specified, if we should use OS copy vs. the git
 // binary, and if a frag/ref has a bad format
-func ParseFile(source string) (details *FileProtoDetails, mods *URLMods) {
+func ParseFile(source string) (details *FileProtoDetails, mods *URLMods, err error) {
+	// NOTE: Why don't we just trim the source here?
 	protoSpecified := false
 	if strings.HasPrefix(source, "file://") && len(source) > 7 {
 		protoSpecified = true
 	}
 
+	refSpecified := false
+	path, ref := "", ""
+	if strings.LastIndex(source, "#") != -1 {
+		refSpecified = true
+
+		segments := strings.SplitN(source, "#", 2)
+		path = segments[0]
+		ref = segments[1]
+	} else {
+		path = source
+	}
+
 	// in each valid case, like the prior logic in scm.go did, we'll make the
 	// paths absolute and prepend file:// to the path which callers should
 	// switch to
 
-	// if source, minus potential file:// prefix and #ref suffix,
-	// exists as is, denote and return
-	if doesExist(source) || doesExist(strings.SplitN(source, "#", 2)[0]) {
-		details = &FileProtoDetails{
-			FileExists:     true,
-			UseCopy:        useCopy(protoSpecified, source),
-			ProtoSpecified: protoSpecified,
-			BadRef:         false,
-		}
-		mods = &URLMods{
-			Scheme: "file",
-			Path:   makePathAbsolute(strings.TrimPrefix(source, "file://")),
-		}
-		return
-	}
-
-	// need to see if this was a file://<valid file>#ref
-	if protoSpecified && strings.LastIndex(source, "#") != -1 {
-		segments := strings.SplitN(source, "#", 2)
-		// given last index check above, segement should
-		// be of len 2
-		path, ref := segments[0], segments[1]
-		// file does not exist, return bad
-		if !doesExist(path) {
+	if doesExist(path) {
+		var isValidGit bool
+		isValidGit, err = isValidGitRepository(path)
+		if err != nil {
 			details = &FileProtoDetails{
 				UseCopy:        false,
 				FileExists:     false,
-				ProtoSpecified: protoSpecified,
 				BadRef:         false,
+				ProtoSpecified: protoSpecified,
 			}
 			mods = nil
 			return
 		}
-		// if ref/frag bad, return bad
-		if !gitSSHURLPathRef.MatchString(ref) {
+
+		if refSpecified {
```

Comment:
i don't see the purpose of this new if statement.  if the ref is "" it will always be valid anyway, so the existing logic should work, no?